tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/AstNode.cpp

```diff
@@ -1318,10 +1310,62 @@ static std::string GetNavigationId(clang::FunctionDecl const* func)
   return navigationId;
 }
 
+static std::string GetNavigationId(clang::CXXRecordDecl const* record)
+{
+  return record->getQualifiedNameAsString();
+}
+
+/** 
+* Get the navigation ID for a qualified type declaration.
+* 
+* The navigation ID uniquely
+ * identifies a declaration in the ApiView. For qualified type
+* declarations, we want to include
+ * the full type name, including the namespace.
+* 
+* @param type The type to get the navigation ID
+ * for.
+* @return The navigation ID for the type.
+* 
+*/
+static std::string GetNavigationId(clang::QualType const& type)
+{
+  std::string navigationId;
+  clang::PrintingPolicy pp{LangOptions{}};
+  pp.adjustForCPlusPlus();
+  pp.FullyQualifiedName = true;
+  pp.TerseOutput = true;
+  llvm::raw_string_ostream os(navigationId);
+
+  type.print(os, pp);
+  return navigationId;
+}
+
+/** 
+* Get the navigation ID for a friend declaration.
+* 
+* The navigation ID uniquely identifies a
+ * declaration in the ApiView. For friend declarations, we
+* want to include the navigation ID of
+ * the parent node, followed by "_friend_", followed by the
+* navigation ID of the friend
+ * declaration.
+* 
+* @param friendDecl The friend declaration to get the navigation ID for.
+* @param
+ * parentNodeNavigationId The navigation ID of the parent node.
+* @return The navigation ID for the
+ * friend declaration.
+* 
+*/
 static std::string GetNavigationId(
     clang::FriendDecl const* friendDecl,
     std::string const& parentNodeNavigationId)
 {
+  if (parentNodeNavigationId.empty())
+  {
+    llvm::errs() << "Parent node navigation ID is empty\n";
```

Comment:
Should we continue after this error?

Reply:
No, I don't think so - this is mostly a diagnostic for me - I've not seen it happen yet but it could and it would result in a weird "friend" navigation name in the apiview - that's not shown, but it could be the cause of duplicate navigation ids.